commit 76c182ed2df4a2f5521b5c1e465fe2d8974f9ee9
Author: Mark Thomas <markt@apache.org>
Date:   Wed Jun 10 21:54:24 2015 +0000

    Sync and notify on the correct object (the Stream) for stream flow control.
    
    git-svn-id: https://svn.apache.org/repos/asf/tomcat/trunk@1684770 13f79535-47bb-0310-9956-ffa450edef68

diff --git a/java/org/apache/coyote/http2/Stream.java b/java/org/apache/coyote/http2/Stream.java
index 492b672..fde119a 100644
--- a/java/org/apache/coyote/http2/Stream.java
+++ b/java/org/apache/coyote/http2/Stream.java
@@ -318,9 +318,9 @@ public class Stream extends AbstractStream implements HeaderEmitter {
                     if (thisWriteStream < 1) {
                         // Need to block until a WindowUpdate message is
                         // processed for this stream
-                        synchronized (this) {
+                        synchronized (Stream.this) {
                             try {
-                                wait();
+                                Stream.this.wait();
                             } catch (InterruptedException e) {
                                 // TODO: Possible shutdown?
                             }
@@ -335,9 +335,9 @@ public class Stream extends AbstractStream implements HeaderEmitter {
                     if (thisWrite < 1) {
                         // Need to block until a WindowUpdate message is
                         // processed for this connection
-                        synchronized (this) {
+                        synchronized (Stream.this) {
                             try {
-                                wait();
+                                Stream.this.wait();
                             } catch (InterruptedException e) {
                                 // TODO: Possible shutdown?
                             }
